package/setup.py

```diff
@@ -130,33 +129,93 @@ def get(self, option_name, default=None):
         except configparser.NoOptionError:
             return default
 
-class MDAExtension(Extension, object):
-    """Derived class to cleanly handle setup-time (numpy) dependencies.
+class MDADistribution(Distribution, object):
+    """Derived class to cleanly handle setup-time (cython/numpy) dependencies.
     """
-    # The only setup-time numpy dependency comes when setting up its
-    #  include dir.
-    # The actual numpy import and call can be delayed until after pip
-    #  has figured it must install numpy.
-    # This is accomplished by passing the get_numpy_include function
-    #  as one of the include_dirs. This derived Extension class takes
-    #  care of calling it when needed.
-    def __init__(self, *args, **kwargs):
-        self._mda_include_dirs = []
-        super(MDAExtension, self).__init__(*args, **kwargs)
-
-    @property
-    def include_dirs(self):
-        if not self._mda_include_dirs:
-            for item in self._mda_include_dir_args:
-                try:
-                    self._mda_include_dirs.append(item()) #The numpy callable
-                except TypeError:
-                    self._mda_include_dirs.append(item)
-        return self._mda_include_dirs
-
-    @include_dirs.setter
-    def include_dirs(self, val):
-        self._mda_include_dir_args = val
+    # Setup-time cython dependency comes when cythonizing sources (cython might
+    #  not be available) or configuring numpy's include dir (numpy might
+    #  not be available either).
+    # The actual cython/numpy imports and calls can be delayed until after
+    #  pip/setuptools have figured they must install them.
+    # This is accomplished by using a distutils.core.Distribution subclass
+    #  that lazily fills-in cython and numpy information by hooking into
+    #  the run_commands method.
+    def __init__(self, attrs=None):
+        # Attributes must exist if they're to be imported from attrs.
+        self.mda_use_cython = is_dev
+        self.mda_build_requires = None
+        super(MDADistribution, self).__init__(attrs)
+
+    def run_commands(self):
+        # The following seems fragile...
```

Comment:
@richardjgowers: I approached the `biopython` people back then, proposing my workaround. They have extremely exquisite deployment targets and it'd have been too much work for me to fix it there. I refactored my workaround a bit now and might try looking at it again, but, honestly, don't hold your breath.

@kain88-de: I disagree that we're doing anything bad here. If you look at the `setuptools` code you'll quickly realize everything feels like a stack of kludges and hacks on top of `distutils`. By subclassing and putting those in our `setup.py` at least the hacks are obvious. If, however, you feel that the code lacks commenting, let me know.

Besides, relying on `setup_requires` is bad (there's no control whatsoever over its behavior), and that's what we already do before this PR! If you don't have `numpy` and run `./setup.py --help` it will install `numpy` in your source dir(!).

To address your question:
(assuming that we invoke `./setup.py install` and that `numpy` is absent)
- Before this PR:
 1. `setup()` is invoked with all our options;
 2. A `setuptools.Distribution` is instanced with those options;
 3. The distribution's `__init__` checks for `setup_requires`, and any needed packages installed in the current dir (under `.eggs`);
 4. `MDAnalysis` gets built (I'd imagine `setuptools` would install dependencies under `install_requires` before building `MDAnalysis`, but apparently that's not the case);
 5. Everything under `install_requires` that isn't available in the system gets system-wide installed. Note that having a `numpy` egg in the current dir inhibits any further system-wide installation of it (see `setuptools`' issues [209](https://bitbucket.org/pypa/setuptools/issues/209/setup_requires-and-install_requires-dont) and [391](https://bitbucket.org/pypa/setuptools/issues/391/dependencies-listed-in-both-setup_requires)).
- After this PR:
 1. `setup()` is invoked with all our options;
 2. Our own subclass of `setuptools.Distribution` is instanced with those options;
 3. The distribution's `__init__` checks for `setup_requires`, but none is present because we now have a better way (I'd really advocate against ever using it again);
 4. Depending on which command is run, we go through `mda_build_requires` and install the required dependencies. This is only done if we detect that we're doing an install of `MDAnalysis`. Asking for `--help` or similar won't trigger dependency installation. 
 5. Our subclassing replaces the dumb local install procedure with a system-wide install (we actually copy the flags passed to `setup.py`, so `--user` installs should also work as expected).
 6. `MDAnalysis` gets built according to the regular `setuptools` flow.

When going through `pip`, it first probes `MDAnalysis` for dependencies, and, recursively, its dependencies. It internally invokes `./setup.py egg_info`, and later `./setup.py install`. `MDAnalysis` only gets built at the install phase, but `numpy` will be installed upon initial probing. Again, this is to keep `biopython` happy; we're already pretty much independent of setup-time dependency solving.